File: foreign/cpp/sdk/net/address.cc
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.
#include "address.h"

const icp::net::protocol::ProtocolDefinition& icp::net::address::LogicalAddress::getProtocolDefinition() const {
    return this->protocolProvider->getProtocolDefinition(this->getProtocol());
}

icp::net::address::LogicalAddress::LogicalAddress(const std::string& url, const icp::net::protocol::ProtocolProvider* protocolProvider) {
    auto parse_result = ada::parse<ada::url>(url);
    if (!parse_result) {
        throw std::invalid_argument("Invalid URL: " + url);
    }
    auto value = parse_result.value();
    auto protocol = value.get_protocol();
    protocol = protocol.substr(0, protocol.length() - 1);
    if (!protocolProvider->isSupported(protocol)) {
        throw std::invalid_argument("Unsupported protocol: " + protocol);
    }
    this->url = value;
    this->protocolProvider = protocolProvider;
}

const uint16_t icp::net::address::LogicalAddress::getPort() const {
    if (url.get_port().empty()) {
        return this->getProtocolDefinition().getDefaultPort();
    } else {
        int port = std::stoi(url.get_port());

        // this should not happen if ada::parse is working correctly
        if (port < 0 || port > 65535) {
            throw std::out_of_range("Port number out of range");
        }

        return port;
    }
}

File: core/server/src/streaming/utils/address.rs
/* Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

/// Extracts IP from an address string like "127.0.0.1:8090" or "[::1]:8090"
pub fn extract_ip(address: &str) -> String {
    if let Some(colon_pos) = address.rfind(':') {
        // Handle IPv6 addresses like [::1]:8090
        if address.starts_with('[')
            && let Some(bracket_pos) = address.rfind(']')
        {
            return address[1..bracket_pos].to_string();
        }
        // Handle IPv4 addresses like 127.0.0.1:8090
        return address[..colon_pos].to_string();
    }
    address.to_string()
}

/// Extracts port from an address string like "127.0.0.1:8090"
pub fn extract_port(address: &str) -> u16 {
    if let Some(colon_pos) = address.rfind(':')
        && let Ok(port) = address[colon_pos + 1..].parse::<u16>()
    {
        return port;
    }
    0
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_extract_ip_ipv4() {
        assert_eq!(extract_ip("127.0.0.1:8090"), "127.0.0.1");
        assert_eq!(extract_ip("192.168.1.100:3000"), "192.168.1.100");
    }

    #[test]
    fn test_extract_ip_ipv6() {
        assert_eq!(extract_ip("[::1]:8090"), "::1");
        assert_eq!(extract_ip("[2001:db8::1]:443"), "2001:db8::1");
    }

    #[test]
    fn test_extract_ip_no_port() {
        assert_eq!(extract_ip("127.0.0.1"), "127.0.0.1");
    }

    #[test]
    fn test_extract_port() {
        assert_eq!(extract_port("127.0.0.1:8090"), 8090);
        assert_eq!(extract_port("192.168.1.100:3000"), 3000);
        assert_eq!(extract_port("[::1]:8090"), 8090);
    }

    #[test]
    fn test_extract_port_no_port() {
        assert_eq!(extract_port("127.0.0.1"), 0);
    }
}
